README.md
# suncalc-py

<p>
  <a href="https://github.com/kylebarron/suncalc-py/actions?query=workflow%3ACI" target="_blank">
      <img src="https://github.com/kylebarron/suncalc-py/workflows/test/badge.svg" alt="Test">
  </a>
  <a href="https://pypi.org/project/suncalc" target="_blank">
      <img src="https://img.shields.io/pypi/v/suncalc?color=%2334D058&label=pypi%20package" alt="Package version">
  </a>
  <a href="https://github.com/kylebarron/suncalc-py/blob/master/LICENSE" target="_blank">
      <img src="https://img.shields.io/github/license/kylebarron/suncalc-py.svg" alt="Downloads">
  </a>
</p>


A fast, vectorized Python implementation of [`suncalc.js`][suncalc-js] for
calculating sun position and sunlight phases (times for sunrise, sunset, dusk,
etc.) for the given location and time.

[suncalc-js]: https://github.com/mourner/suncalc

While other similar libraries exist, I didn't originally encounter any that met
my requirements of being openly-licensed, vectorized, and simple to use <sup>1</sup>.

## Install

```
pip install suncalc
```

## Using

### Example

`suncalc` is designed to work both with single values and with arrays of values.

First, import the module:

```py
from suncalc import get_position, get_times
from datetime import datetime
```

There are currently two methods: `get_position`, to get the sun azimuth and
altitude for a given date and position, and `get_times`, to get sunlight phases
for a given date and position.

```py
date = datetime.now()
lon = 20
lat = 45
get_position(date, lon, lat)
# {'azimuth': -0.8619668996997687, 'altitude': 0.5586446727994595}

get_times(date, lon, lat)
# {'solar_noon': Timestamp('2020-11-20 08:47:08.410863770'),
#  'nadir': Timestamp('2020-11-19 20:47:08.410863770'),
#  'sunrise': Timestamp('2020-11-20 03:13:22.645455322'),
#  'sunset': Timestamp('2020-11-20 14:20:54.176272461'),
#  'sunrise_end': Timestamp('2020-11-20 03:15:48.318936035'),
#  'sunset_start': Timestamp('2020-11-20 14:18:28.502791748'),
#  'dawn': Timestamp('2020-11-20 02:50:00.045539551'),
#  'dusk': Timestamp('2020-11-20 14:44:16.776188232'),
#  'nautical_dawn': Timestamp('2020-11-20 02:23:10.019832520'),
#  'nautical_dusk': Timestamp('2020-11-20 15:11:06.801895264'),
#  'night_end': Timestamp('2020-11-20 01:56:36.144269287'),
#  'night': Timestamp('2020-11-20 15:37:40.677458252'),
#  'golden_hour_end': Timestamp('2020-11-20 03:44:46.795967773'),
#  'golden_hour': Timestamp('2020-11-20 13:49:30.025760010')}
```

These methods also work for _arrays_ of data, and since the implementation is
vectorized it's much faster than a for loop in Python.

```py
import pandas as pd

df = pd.DataFrame({
    'date': [date] * 10,
    'lon': [lon] * 10,
    'lat': [lat] * 10
})
pd.DataFrame(get_position(df['date'], df['lon'], df['lat']))
# azimuth	altitude
# 0	-1.485509	-1.048223
# 1	-1.485509	-1.048223
# ...

pd.DataFrame(get_times(df['date'], df['lon'], df['lat']))['solar_noon']
# 0   2020-11-20 08:47:08.410863872+00:00
# 1   2020-11-20 08:47:08.410863872+00:00
# ...
# Name: solar_noon, dtype: datetime64[ns, UTC]
```

If you want to join this data back to your `DataFrame`, you can use `pd.concat`:

```py
times = pd.DataFrame(get_times(df['date'], df['lon'], df['lat']))
pd.concat([df, times], axis=1)
```

### API

#### `get_position`

Calculate sun position (azimuth and altitude) for a given date and
latitude/longitude

- `date` (`datetime` or a pandas series of datetimes): date and time to find sun position of. **Datetime must be in UTC**.
- `lng` (`float` or numpy array of `float`): longitude to find sun position of
- `lat` (`float` or numpy array of `float`): latitude to find sun position of

Returns a `dict` with two keys: `azimuth` and `altitude`. If the input values
were singletons, the `dict`'s values will be floats. Otherwise they'll be numpy
arrays of floats.

#### `get_times`

- `date` (`datetime` or a pandas series of datetimes): date and time to find sunlight phases of. **Datetime must be in UTC**.
- `lng` (`float` or numpy array of `float`): longitude to find sunlight phases of
- `lat` (`float` or numpy array of `float`): latitude to find sunlight phases of
- `height` (`float` or numpy array of `float`, default `0`): observer height in meters
- `times` (`Iterable[Tuple[float, str, str]]`): an iterable defining the angle above the horizon and strings for custom sunlight phases. The default is:

    ```py
    # (angle, morning name, evening name)
    DEFAULT_TIMES = [
        (-0.833, 'sunrise', 'sunset'),
        (-0.3, 'sunrise_end', 'sunset_start'),
        (-6, 'dawn', 'dusk'),
        (-12, 'nautical_dawn', 'nautical_dusk'),
        (-18, 'night_end', 'night'),
        (6, 'golden_hour_end', 'golden_hour')
    ]
    ```

Returns a `dict` where the keys are `solar_noon`, `nadir`, plus any keys passed
in the `times` argument. If the input values were singletons, the `dict`'s
values will be of type `datetime.datetime` (or `pd.Timestamp` if you have pandas
installed, which is a subclass of and therefore compatible with
`datetime.datetime`). Otherwise they'll be pandas `DateTime` series. **The
returned times will be in UTC.**

## Benchmark

This benchmark is to show that the vectorized implementation is nearly 100x
faster than a for loop in Python.

First set up a `DataFrame` with random data. Here I create 100,000 rows.

```py
from suncalc import get_position, get_times
import pandas as pd

def random_dates(start, end, n=10):
    """Create an array of random dates"""
    start_u = start.value//10**9
    end_u = end.value//10**9
    return pd.to_datetime(np.random.randint(start_u, end_u, n), unit='s')

start = pd.to_datetime('2015-01-01')
end = pd.to_datetime('2018-01-01')
dates = random_dates(start, end, n=100_000)

lons = np.random.uniform(low=-179, high=179, size=(100_000,))
lats = np.random.uniform(low=-89, high=89, size=(100_000,))

df = pd.DataFrame({'date': dates, 'lat': lats, 'lon': lons})
```

Then compute `SunCalc.get_position` two ways: the first using the vectorized
implementation and the second using `df.apply`, which is equivalent to a for
loop. The first is more than **100x faster** than the second.

```py
%timeit get_position(df['date'], df['lon'], df['lat'])
# 41.4 ms ± 437 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit df.apply(lambda row: get_position(row['date'], row['lon'], row['lat']), axis=1)
# 4.89 s ± 184 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Likewise, compute `SunCalc.get_times` the same two ways: first using the
vectorized implementation and the second using `df.apply`. The first is **2800x
faster** than the second! Some of the difference here is that under the hood the
non-vectorized approach uses `pd.to_datetime` while the vectorized
implementation uses `np.astype('datetime64[ns, UTC]')`. `pd.to_datetime` is
really slow!!

```py
%timeit get_times(df['date'], df['lon'], df['lat'])
# 55.3 ms ± 1.91 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%time df.apply(lambda row: get_times(row['date'], row['lon'], row['lat']), axis=1)
# CPU times: user 2min 33s, sys: 288 ms, total: 2min 34s
# Wall time: 2min 34s
```

---

1: [`pyorbital`](https://github.com/pytroll/pyorbital) looks great but is
GPL3-licensed; [`pysolar`](https://github.com/pingswept/pysolar) is also
GPL3-licensed; [`pyEphem`](https://rhodesmill.org/pyephem/) is LGPL3-licensed.
[`suncalcPy`](https://github.com/Broham/suncalcPy) is another port of
`suncalc.js`, and is MIT-licensed, but doesn't use Numpy and thus isn't
vectorized. I recently discovered [`sunpy`](https://github.com/sunpy/sunpy) and
[`astropy`](https://github.com/astropy/astropy), both of which probably would've
worked but I didn't see them at first and they look quite complex for this
simple task...


CHANGELOG.md
# Changelog

## [0.1.3] - 2023-04-18

- Ensure pandas 2.0 compatibility (fix integer casting of datetimes)

## [0.1.2] - 2020-12-02

- Try to catch NaN before passing to `datetime.utcfromtimestamp`

## [0.1.1] - 2020-11-20

- Fix PyPI install by adding `MANIFEST.in`
- Update documentation

## [0.1.0] - 2020-11-19

- Initial release on PyPI


tests/test_suncalc.py
from datetime import datetime, timezone

import numpy as np
import pandas as pd

from suncalc import get_position, get_times

date = datetime(2013, 3, 5, tzinfo=timezone.utc)
lat = 50.5
lng = 30.5
height = 2000

testTimes = {
    'solar_noon': '2013-03-05T10:10:57Z',
    'nadir': '2013-03-04T22:10:57Z',
    'sunrise': '2013-03-05T04:34:56Z',
    'sunset': '2013-03-05T15:46:57Z',
    'sunrise_end': '2013-03-05T04:38:19Z',
    'sunset_start': '2013-03-05T15:43:34Z',
    'dawn': '2013-03-05T04:02:17Z',
    'dusk': '2013-03-05T16:19:36Z',
    'nautical_dawn': '2013-03-05T03:24:31Z',
    'nautical_dusk': '2013-03-05T16:57:22Z',
    'night_end': '2013-03-05T02:46:17Z',
    'night': '2013-03-05T17:35:36Z',
    'golden_hour_end': '2013-03-05T05:19:01Z',
    'golden_hour': '2013-03-05T15:02:52Z'}

heightTestTimes = {
    'solar_noon': '2013-03-05T10:10:57Z',
    'nadir': '2013-03-04T22:10:57Z',
    'sunrise': '2013-03-05T04:25:07Z',
    'sunset': '2013-03-05T15:56:46Z'}


def test_get_position():
    """getPosition returns azimuth and altitude for the given time and location
    """
    pos = get_position(date, lng, lat)
    assert np.isclose(pos['azimuth'], -2.5003175907168385)
    assert np.isclose(pos['altitude'], -0.7000406838781611)


def test_get_times():
    """getTimes returns sun phases for the given date and location
    """
    times = get_times(date, lng, lat)
    for key, value in testTimes.items():
        assert times[key].strftime("%Y-%m-%dT%H:%M:%SZ") == value


def test_get_times_height():
    """getTimes adjusts sun phases when additionally given the observer height
    """
    times = get_times(date, lng, lat, height)
    for key, value in heightTestTimes.items():
        assert times[key].strftime("%Y-%m-%dT%H:%M:%SZ") == value


def test_get_position_pandas_single_timestamp():
    ts_date = pd.Timestamp(date)

    pos = get_position(ts_date, lng, lat)
    assert np.isclose(pos['azimuth'], -2.5003175907168385)
    assert np.isclose(pos['altitude'], -0.7000406838781611)


def test_get_position_pandas_datetime_series():
    df = pd.DataFrame({'date': [date] * 3, 'lat': [lat] * 3, 'lng': [lng] * 3})

    pos = pd.DataFrame(get_position(df['date'], df['lng'], df['lat']))

    assert pos.shape == (3, 2)
    assert all(x in pos.columns for x in ['azimuth', 'altitude'])
    assert pos.dtypes['azimuth'] == np.dtype('float64')
    assert pos.dtypes['altitude'] == np.dtype('float64')

    assert np.isclose(pos['azimuth'].iloc[0], -2.5003175907168385)
    assert np.isclose(pos['altitude'].iloc[0], -0.7000406838781611)


def test_get_times_pandas_single():
    times = get_times(date, lng, lat)

    assert isinstance(times['solar_noon'], pd.Timestamp)


def test_get_times_datetime_single():
    times = get_times(date, lng, lat)

    # This is true because pd.Timestamp is an instance of datetime.datetime
    assert isinstance(times['solar_noon'], datetime)


def test_get_times_arrays():
    df = pd.DataFrame({'date': [date] * 3, 'lat': [lat] * 3, 'lng': [lng] * 3})

    times = pd.DataFrame(get_times(df['date'], df['lng'], df['lat']))

    assert pd.api.types.is_datetime64_any_dtype(times['solar_noon'])

    assert times['solar_noon'].iloc[0].strftime(
        "%Y-%m-%dT%H:%M:%SZ") == testTimes['solar_noon']


def test_get_times_for_high_latitudes():
    """getTimes may fail (maybe only on Windows?) for high latitudes in the summer

    See https://github.com/kylebarron/suncalc-py/issues/4
    """
    date = datetime(2020, 5, 26, 0, 0, 0)
    lng = -114.0719
    lat = 51.0447

    # Make sure this doesn't raise an exception (though it will emit a warning
    # due to a division error)
    times = get_times(date, lng, lat)


# t.test('getMoonPosition returns moon position data given time and location', function (t) {
#     var moonPos = SunCalc.getMoonPosition(date, lng, lat);
#
#     t.ok(near(moonPos.azimuth, -0.9783999522438226), 'azimuth');
#     t.ok(near(moonPos.altitude, 0.014551482243892251), 'altitude');
#     t.ok(near(moonPos.distance, 364121.37256256194), 'distance');
#     t.end();
# });
#
# t.test('getMoonIllumination returns fraction and angle of moon\'s illuminated limb and phase', function (t) {
#     var moonIllum = SunCalc.getMoonIllumination(date);
#
#     t.ok(near(moonIllum.fraction, 0.4848068202456373), 'fraction');
#     t.ok(near(moonIllum.phase, 0.7548368838538762), 'phase');
#     t.ok(near(moonIllum.angle, 1.6732942678578346), 'angle');
#     t.end();
# });
#
# t.test('getMoonTimes returns moon rise and set times', function (t) {
#     var moonTimes = SunCalc.getMoonTimes(new Date('2013-03-04UTC'), lng, lat, true);
#
#     t.equal(moonTimes.rise.toUTCString(), 'Mon, 04 Mar 2013 23:54:29 GMT');
#     t.equal(moonTimes.set.toUTCString(), 'Mon, 04 Mar 2013 07:47:58 GMT');
#
#     t.end();
# });


suncalc/suncalc.py
"""
suncalc-py is ported from suncalc.js under the BSD-2-Clause license.

Copyright (c) 2014, Vladimir Agafonkin
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are
permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, this list of
      conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above copyright notice, this list
      of conditions and the following disclaimer in the documentation and/or other materials
      provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

from datetime import datetime
from typing import Iterable, Tuple

import numpy as np

try:
    import pandas as pd
except ImportError:
    pd = None

# shortcuts for easier to read formulas
PI = np.pi
sin = np.sin
cos = np.cos
tan = np.tan
asin = np.arcsin
atan = np.arctan2
acos = np.arccos
rad = PI / 180

# sun times configuration (angle, morning name, evening name)
DEFAULT_TIMES = [
    (-0.833, 'sunrise', 'sunset'),
    (-0.3, 'sunrise_end', 'sunset_start'),
    (-6, 'dawn', 'dusk'),
    (-12, 'nautical_dawn', 'nautical_dusk'),
    (-18, 'night_end', 'night'),
    (6, 'golden_hour_end', 'golden_hour')
] # yapf: disable

# date/time constants and conversions
dayMs = 1000 * 60 * 60 * 24
J1970 = 2440588
J2000 = 2451545


def to_milliseconds(date):
    # datetime.datetime
    if isinstance(date, datetime):
        return date.timestamp() * 1000

    # Pandas series of Pandas datetime objects
    if pd and pd.api.types.is_datetime64_any_dtype(date):
        # A datetime-like series coerce to int is (always?) in nanoseconds
        return date.astype('int64') / 10 ** 6

    # Single pandas Timestamp
    if pd and isinstance(date, pd.Timestamp):
        date = date.to_numpy()

    # Numpy datetime64
    if np.issubdtype(date.dtype, np.datetime64):
        return date.astype('datetime64[ms]').astype('int64')

    # Last-ditch effort
    if pd:
        return np.array(pd.to_datetime(date).astype('int64') / 10 ** 6)

    raise ValueError(f'Unknown date type: {type(date)}')


def to_julian(date):
    return to_milliseconds(date) / dayMs - 0.5 + J1970


def from_julian(j):
    ms_date = (j + 0.5 - J1970) * dayMs

    if pd:
        # If a single value, coerce to a pd.Timestamp
        if np.prod(np.array(ms_date).shape) == 1:
            return pd.to_datetime(ms_date, unit='ms')

        # .astype(datetime) is much faster than pd.to_datetime but it only works
        # on series of dates, not on a single pd.Timestamp, so I fall back to
        # pd.to_datetime for that.
        try:
            return (pd.Series(ms_date) * 1e6).astype('datetime64[ns, UTC]')
        except TypeError:
            return pd.to_datetime(ms_date, unit='ms')

    # ms_date could be iterable
    try:
        return np.array([
            datetime.utcfromtimestamp(x / 1000)
            if not np.isnan(x) else np.datetime64('NaT') for x in ms_date])

    except TypeError:
        return datetime.utcfromtimestamp(
            ms_date / 1000) if not np.isnan(ms_date) else np.datetime64('NaT')


def to_days(date):
    return to_julian(date) - J2000


# general calculations for position

# obliquity of the Earth
e = rad * 23.4397


def right_ascension(l, b):
    return atan(sin(l) * cos(e) - tan(b) * sin(e), cos(l))


def declination(l, b):
    return asin(sin(b) * cos(e) + cos(b) * sin(e) * sin(l))


def azimuth(H, phi, dec):
    return atan(sin(H), cos(H) * sin(phi) - tan(dec) * cos(phi))


def altitude(H, phi, dec):
    return asin(sin(phi) * sin(dec) + cos(phi) * cos(dec) * cos(H))


def sidereal_time(d, lw):
    return rad * (280.16 + 360.9856235 * d) - lw


def astro_refraction(h):
    # the following formula works for positive altitudes only.
    # if h = -0.08901179 a div/0 would occur.
    h = np.maximum(h, 0)

    # formula 16.4 of "Astronomical Algorithms" 2nd edition by Jean Meeus
    # (Willmann-Bell, Richmond) 1998. 1.02 / tan(h + 10.26 / (h + 5.10)) h in
    # degrees, result in arc minutes -> converted to rad:
    return 0.0002967 / np.tan(h + 0.00312536 / (h + 0.08901179))


# general sun calculations


def solar_mean_anomaly(d):
    return rad * (357.5291 + 0.98560028 * d)


def ecliptic_longitude(M):
    # equation of center
    C = rad * (1.9148 * sin(M) + 0.02 * sin(2 * M) + 0.0003 * sin(3 * M))

    # perihelion of the Earth
    P = rad * 102.9372

    return M + C + P + PI


def sun_coords(d):
    M = solar_mean_anomaly(d)
    L = ecliptic_longitude(M)

    return {'dec': declination(L, 0), 'ra': right_ascension(L, 0)}


# calculations for sun times
J0 = 0.0009


def julian_cycle(d, lw):
    return np.round(d - J0 - lw / (2 * PI))


def approx_transit(Ht, lw, n):
    return J0 + (Ht + lw) / (2 * PI) + n


def solar_transit_j(ds, M, L):
    return J2000 + ds + 0.0053 * sin(M) - 0.0069 * sin(2 * L)


def hour_angle(h, phi, d):
    return acos((sin(h) - sin(phi) * sin(d)) / (cos(phi) * cos(d)))


def observer_angle(height):
    return -2.076 * np.sqrt(height) / 60


def get_set_j(h, lw, phi, dec, n, M, L):
    """Get set time for the given sun altitude
    """
    w = hour_angle(h, phi, dec)
    a = approx_transit(w, lw, n)
    return solar_transit_j(a, M, L)


def get_position(date, lng, lat):
    """Calculate sun position for a given date and latitude/longitude
    """
    lw = rad * -lng
    phi = rad * lat
    d = to_days(date)

    c = sun_coords(d)
    H = sidereal_time(d, lw) - c['ra']

    return {
        'azimuth': azimuth(H, phi, c['dec']),
        'altitude': altitude(H, phi, c['dec'])}


def get_times(
        date,
        lng,
        lat,
        height=0,
        times: Iterable[Tuple[float, str, str]] = DEFAULT_TIMES):
    """Calculate sun times

    Calculate sun times for a given date, latitude/longitude, and,
    optionally, the observer height (in meters) relative to the horizon
    """
    # If inputs are vectors (or some list-like type), then coerce them to
    # numpy arrays
    #
    # When inputs are pandas series, then intermediate objects will also be
    # pandas series, and you won't be able to do 2d broadcasting.
    try:
        len(date)
        len(lat)
        len(lng)
        array_input = True
        date = np.array(date)
        lat = np.array(lat)
        lng = np.array(lng)
    except TypeError:
        array_input = False

    lw = rad * -lng
    phi = rad * lat

    dh = observer_angle(height)

    d = to_days(date)
    n = julian_cycle(d, lw)
    ds = approx_transit(0, lw, n)

    M = solar_mean_anomaly(ds)
    L = ecliptic_longitude(M)
    dec = declination(L, 0)

    Jnoon = solar_transit_j(ds, M, L)

    result = {
        'solar_noon': from_julian(Jnoon),
        'nadir': from_julian(Jnoon - 0.5)}

    angles = np.array([time[0] for time in times])
    h0 = (angles + dh) * rad

    # If array input, add an axis to allow 2d broadcasting
    if array_input:
        h0 = h0[:, np.newaxis]

    # Need to add an axis for 2D broadcasting
    Jset = get_set_j(h0, lw, phi, dec, n, M, L)
    Jrise = Jnoon - (Jset - Jnoon)

    for idx, time in enumerate(times):
        if array_input:
            result[time[1]] = from_julian(Jrise[idx, :])
            result[time[2]] = from_julian(Jset[idx, :])
        else:
            result[time[1]] = from_julian(Jrise[idx])
            result[time[2]] = from_julian(Jset[idx])

    return result


# moon calculations, based on http://aa.quae.nl/en/reken/hemelpositie.html
# formulas


def moon_coords(d):
    """Geocentric ecliptic coordinates of the moon
    """

    # ecliptic longitude
    L = rad * (218.316 + 13.176396 * d)
    # mean anomaly
    M = rad * (134.963 + 13.064993 * d)
    # mean distance
    F = rad * (93.272 + 13.229350 * d)

    # longitude
    l = L + rad * 6.289 * sin(M)
    # latitude
    b = rad * 5.128 * sin(F)
    # distance to the moon in km
    dt = 385001 - 20905 * cos(M)

    return {'ra': right_ascension(l, b), 'dec': declination(l, b), 'dist': dt}


def getMoonPosition(date, lat, lng):

    lw = rad * -lng
    phi = rad * lat
    d = to_days(date)

    c = moon_coords(d)
    H = sidereal_time(d, lw) - c['ra']
    h = altitude(H, phi, c['dec'])

    # formula 14.1 of "Astronomical Algorithms" 2nd edition by Jean Meeus
    # (Willmann-Bell, Richmond) 1998.
    pa = atan(sin(H), tan(phi) * cos(c['dec']) - sin(c['dec']) * cos(H))

    # altitude correction for refraction
    h = h + astro_refraction(h)

    return {
        'azimuth': azimuth(H, phi, c['dec']),
        'altitude': h,
        'distance': c['dist'],
        'parallacticAngle': pa}


# calculations for illumination parameters of the moon, based on
# http://idlastro.gsfc.nasa.gov/ftp/pro/astro/mphase.pro formulas and Chapter 48
# of "Astronomical Algorithms" 2nd edition by Jean Meeus (Willmann-Bell,
# Richmond) 1998.


def getMoonIllumination(date):

    d = to_days(date)
    s = sun_coords(d)
    m = moon_coords(d)

    # distance from Earth to Sun in km
    sdist = 149598000

    phi = acos(
        sin(s['dec']) * sin(m['dec']) +
        cos(s['dec']) * cos(m['dec']) * cos(s['ra'] - m['ra']))
    inc = atan(sdist * sin(phi), m['dist'] - sdist * cos(phi)),
    angle = atan(
        cos(s['dec']) * sin(s['ra'] - m['ra']),
        sin(s['dec']) * cos(m['dec']) -
        cos(s['dec']) * sin(m['dec']) * cos(s['ra'] - m['ra']))

    return {
        'fraction': (1 + cos(inc)) / 2,
        'phase': 0.5 + 0.5 * inc * np.sign(angle) / PI,
        'angle': angle}


# def hoursLater(date, h):
#     # TODO: pythonize
#     return new Date(date.valueOf() + h * dayMs / 24)

# calculations for moon rise/set times are based on
# http://www.stargazing.net/kepler/moonrise.html article

# def getMoonTimes(date, lat, lng, inUTC):
#     var t = new Date(date);
#     if (inUTC) t.setUTCHours(0, 0, 0, 0);
#     else t.setHours(0, 0, 0, 0);
#
#     var hc = 0.133 * rad,
#         h0 = SunCalc.getMoonPosition(t, lat, lng).altitude - hc,
#         h1, h2, rise, set, a, b, xe, ye, d, roots, x1, x2, dx;
#
#     // go in 2-hour chunks, each time seeing if a 3-point quadratic curve crosses zero (which means rise or set)
#     for (var i = 1; i <= 24; i += 2) {
#         h1 = SunCalc.getMoonPosition(hoursLater(t, i), lat, lng).altitude - hc;
#         h2 = SunCalc.getMoonPosition(hoursLater(t, i + 1), lat, lng).altitude - hc;
#
#         a = (h0 + h2) / 2 - h1;
#         b = (h2 - h0) / 2;
#         xe = -b / (2 * a);
#         ye = (a * xe + b) * xe + h1;
#         d = b * b - 4 * a * h1;
#         roots = 0;
#
#         if (d >= 0) {
#             dx = Math.sqrt(d) / (Math.abs(a) * 2);
#             x1 = xe - dx;
#             x2 = xe + dx;
#             if (Math.abs(x1) <= 1) roots++;
#             if (Math.abs(x2) <= 1) roots++;
#             if (x1 < -1) x1 = x2;
#         }
#
#         if (roots === 1) {
#             if (h0 < 0) rise = i + x1;
#             else set = i + x1;
#
#         } else if (roots === 2) {
#             rise = i + (ye < 0 ? x2 : x1);
#             set = i + (ye < 0 ? x1 : x2);
#         }
#
#         if (rise && set) break;
#
#         h0 = h2;
#     }
#
#     var result = {};
#
#     if (rise) result.rise = hoursLater(t, rise);
#     if (set) result.set = hoursLater(t, set);
#
#     if (!rise && !set) result[ye > 0 ? 'alwaysUp' : 'alwaysDown'] = true;
#
#     return result;
# };
